File: bdp-devops/ambari-versions/ambari-3.0.0.0-SNAPSHOT/contrib/views/utils/src/main/java/org/apache/ambari/view/utils/hdfs/HdfsUtil.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ambari.view.utils.hdfs;


import org.apache.ambari.view.ViewContext;
import org.apache.commons.io.IOUtils;
import org.apache.hadoop.fs.FSDataInputStream;
import org.apache.hadoop.fs.FSDataOutputStream;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.security.PrivilegedExceptionAction;
import java.util.Map;

public class HdfsUtil {
  private final static Logger LOG =
      LoggerFactory.getLogger(HdfsUtil.class);

  /**
   * Write string to file with overwriting
   * @param filePath path to file
   * @param content new content of file
   */
  public static void putStringToFile(final HdfsApi hdfs,final String filePath, final String content) throws HdfsApiException {

    try {
      synchronized (hdfs) {
        hdfs.execute(new PrivilegedExceptionAction<Void>() {
          @Override
          public Void run() throws Exception {
            final FSDataOutputStream stream = hdfs.create(filePath, true);
            stream.write(content.getBytes());
            stream.close();
            return null;
          }
        }, true);
      }
    } catch (IOException e) {
      throw new HdfsApiException("HDFS020 Could not write file " + filePath, e);
    } catch (InterruptedException e) {
      throw new HdfsApiException("HDFS021 Could not write file " + filePath, e);
    }
  }

  /**
   * Read string from file
   * @param filePath path to file
   */
  public static String readFile(HdfsApi hdfs, String filePath) throws HdfsApiException {
    FSDataInputStream stream;
    try {
      stream = hdfs.open(filePath);
      return IOUtils.toString(stream);
    } catch (IOException e) {
      throw new HdfsApiException("HDFS060 Could not read file " + filePath, e);
    } catch (InterruptedException e) {
      throw new HdfsApiException("HDFS061 Could not read file " + filePath, e);
    }
  }


  /**
   * Increment index appended to filename until find first unallocated file
   * @param fullPathAndFilename path to file and prefix for filename
   * @param extension file extension
   * @return if fullPathAndFilename="/tmp/file",extension=".txt" then filename will be like "/tmp/file_42.txt"
   */
  public static String findUnallocatedFileName(HdfsApi hdfs, String fullPathAndFilename, String extension)
      throws HdfsApiException {
    int triesCount = 0;
    String newFilePath;
    boolean isUnallocatedFilenameFound;

    try {
      do {
        newFilePath = String.format(fullPathAndFilename + "%s" + extension, (triesCount == 0) ? "" : "_" + triesCount);
        LOG.debug("Trying to find free filename " + newFilePath);

        isUnallocatedFilenameFound = !hdfs.exists(newFilePath);
        if (isUnallocatedFilenameFound) {
          LOG.debug("File created successfully!");
        }

        triesCount += 1;
        if (triesCount > 1000) {
          throw new HdfsApiException("HDFS100 Can't find unallocated file name " + fullPathAndFilename + "...");
        }
      } while (!isUnallocatedFilenameFound);
    } catch (IOException e) {
      throw new HdfsApiException("HDFS030 Error in creation " + fullPathAndFilename + "...", e);
    } catch (InterruptedException e) {
      throw new HdfsApiException("HDFS031 Error in creation " + fullPathAndFilename + "...", e);
    }

    return newFilePath;
  }

  /**
   * takes any custom properties that a view wants to be included into the config
   * @param context
   * @param customViewProperties
   * @return
   * @throws HdfsApiException
   */
  public static synchronized HdfsApi connectToHDFSApi(ViewContext context, Map<String, String> customViewProperties)
    throws HdfsApiException {
    ClassLoader currentClassLoader = Thread.currentThread().getContextClassLoader();
    Thread.currentThread().setContextClassLoader(HdfsUtil.class.getClassLoader());
    try {
      ConfigurationBuilder configurationBuilder = new ConfigurationBuilder(context, customViewProperties);
      return getHdfsApi(context, configurationBuilder);
    } finally {
      Thread.currentThread().setContextClassLoader(currentClassLoader);
    }
  }

  /**
   * Factory of HdfsApi for specific ViewContext
   * @param context ViewContext that contains connection credentials
   * @return HdfsApi object
   */
  public static synchronized HdfsApi connectToHDFSApi(ViewContext context) throws HdfsApiException {
    ClassLoader currentClassLoader = Thread.currentThread().getContextClassLoader();
    Thread.currentThread().setContextClassLoader(HdfsUtil.class.getClassLoader());
    try {
      ConfigurationBuilder configurationBuilder = new ConfigurationBuilder(context);
      return getHdfsApi(context, configurationBuilder);
    } finally {
      Thread.currentThread().setContextClassLoader(currentClassLoader);
    }
  }

  private static HdfsApi getHdfsApi(ViewContext context, ConfigurationBuilder configurationBuilder) throws HdfsApiException {
    HdfsApi api = null;
    AuthConfigurationBuilder authConfigurationBuilder = new AuthConfigurationBuilder(context);
    Map<String, String> authParams = authConfigurationBuilder.build();
    configurationBuilder.setAuthParams(authParams);
    try {
      api = new HdfsApi(configurationBuilder, getHdfsUsername(context));
      LOG.info("HdfsApi connected OK");
    } catch (IOException e) {
      LOG.error("exception occurred while creating hdfsApi objcet : {}", e.getMessage(), e);
      String message = "HDFS040 Couldn't open connection to HDFS";
      LOG.error(message);
      throw new HdfsApiException(message, e);
    } catch (InterruptedException e) {
      LOG.error("exception occurred while creating hdfsApi objcet : {}", e.getMessage(), e);
      String message = "HDFS041 Couldn't open connection to HDFS";
      LOG.error(message);
      throw new HdfsApiException(message, e);
    }
    return api;
  }

  /**
   * Returns username for HdfsApi from "webhdfs.username" property if set,
   * if not set then current Ambari username
   * @param context ViewContext
   * @return username
   */
  public static String getHdfsUsername(ViewContext context) {
    String userName = context.getProperties().get("webhdfs.username");
    if (userName == null || userName.compareTo("null") == 0 || userName.compareTo("") == 0) {
      userName = context.getUsername();
    }
    return userName;
  }
}

File: govern/data-quality/griffin/measure/src/main/scala/org/apache/griffin/measure/utils/HdfsUtil.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.griffin.measure.utils

import org.apache.hadoop.fs.{FSDataInputStream, FSDataOutputStream, Path}

import org.apache.griffin.measure.Loggable

object HdfsUtil extends Loggable {

  private val seprator = "/"

  private def getFS(implicit path: Path) = FSUtil.getFileSystem(path.toString)

  def existPath(filePath: String): Boolean = {
    try {
      implicit val path: Path = new Path(filePath)
      getFS.exists(path)
    } catch {
      case _: Throwable => false
    }
  }

  def existFileInDir(dirPath: String, fileName: String): Boolean = {
    val filePath = getHdfsFilePath(dirPath, fileName)
    existPath(filePath)
  }

  def createFile(filePath: String): FSDataOutputStream = {
    implicit val path: Path = new Path(filePath)
    if (getFS.exists(path)) getFS.delete(path, true)
    getFS.create(path)
  }

  def appendOrCreateFile(filePath: String): FSDataOutputStream = {
    implicit val path: Path = new Path(filePath)
    if (getFS.getConf.getBoolean("dfs.support.append", false) && getFS.exists(path)) {
      getFS.append(path)
    } else createFile(filePath)
  }

  def openFile(filePath: String): FSDataInputStream = {
    implicit val path: Path = new Path(filePath)
    getFS.open(path)
  }

  def writeContent(filePath: String, message: String): Unit = {
    val out = createFile(filePath)
    out.write(message.getBytes("utf-8"))
    out.close()
  }

  def withHdfsFile(filePath: String, appendIfExists: Boolean = true)(
      f: FSDataOutputStream => Unit): Unit = {
    val out =
      if (appendIfExists) {
        appendOrCreateFile(filePath)
      } else {
        createFile(filePath)
      }

    f(out)
    out.close()
  }

  def createEmptyFile(filePath: String): Unit = {
    val out = createFile(filePath)
    out.close()
  }

  def getHdfsFilePath(parentPath: String, fileName: String): String = {
    if (parentPath.endsWith(seprator)) parentPath + fileName else parentPath + seprator + fileName
  }

  def deleteHdfsPath(dirPath: String): Unit = {
    try {
      implicit val path: Path = new Path(dirPath)
      if (getFS.exists(path)) getFS.delete(path, true)
    } catch {
      case e: Throwable => error(s"delete path [$dirPath] error: ${e.getMessage}", e)
    }
  }

  def listSubPathsByType(
      dirPath: String,
      subType: String,
      fullPath: Boolean = false): Iterable[String] = {
    if (existPath(dirPath)) {
      try {
        implicit val path: Path = new Path(dirPath)
        val fileStatusArray = getFS.listStatus(path)
        fileStatusArray
          .filter { fileStatus =>
            subType match {
              case "dir" => fileStatus.isDirectory
              case "file" => fileStatus.isFile
              case _ => true
            }
          }
          .map { fileStatus =>
            val fname = fileStatus.getPath.getName
            if (fullPath) getHdfsFilePath(dirPath, fname) else fname
          }
      } catch {
        case e: Throwable =>
          warn(s"list path [$dirPath] warn: ${e.getMessage}", e)
          Nil
      }
    } else Nil
  }

  def listSubPathsByTypes(
      dirPath: String,
      subTypes: Iterable[String],
      fullPath: Boolean = false): Iterable[String] = {
    subTypes.flatMap { subType =>
      listSubPathsByType(dirPath, subType, fullPath)
    }
  }

  def fileNameFromPath(filePath: String): String = {
    val path = new Path(filePath)
    path.getName
  }
}
